clients/src/main/java/org/apache/kafka/common/requests/ProduceRequest.java

```diff
@@ -51,13 +52,20 @@ public static Builder forMagic(byte magic, ProduceRequestData data) {
         if (magic < RecordBatch.MAGIC_VALUE_V2) {
             minVersion = 2;
             maxVersion = 2;
+        } else if (canNotSupportTopicId(data)) {
+            minVersion = 3;
+            maxVersion = 11;
         } else {
             minVersion = 3;
             maxVersion = ApiKeys.PRODUCE.latestVersion();
         }
         return new Builder(minVersion, maxVersion, data);
     }
 
+    private static boolean canNotSupportTopicId(ProduceRequestData data) {
+        return data.topicData().stream().anyMatch(d -> d.topicId() == null || d.topicId() == Uuid.ZERO_UUID);
```

Comment:
I saw this null check in a few places. Is there a chance that this value is null or are we just trying to cover our bases?

Reply:
just covering our bases, but in this method we will not get null as the ProduceRequest constructor has ZERO_UUID as default. Will tide this up